cluster-autoscaler/core/scaledown/latencytracker/node_latency_tracker.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package latencytracker
+
+import (
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+	ca_context "k8s.io/autoscaler/cluster-autoscaler/context"
+	"k8s.io/autoscaler/cluster-autoscaler/core/scaledown/status"
+	"k8s.io/autoscaler/cluster-autoscaler/metrics"
+	"k8s.io/klog/v2"
+)
+
+type unneededNodeState struct {
+	unneededSince time.Time
+	threshold     time.Duration
+}
+
+// NodeLatencyTracker is a concrete implementation of LatencyTracker.
+// It keeps track of nodes that are marked as unneeded, when they became unneeded,
+// and thresholds to adjust node removal latency metrics.
+type NodeLatencyTracker struct {
+	unneededNodes    map[string]unneededNodeState
+	unremovableNodes map[string]struct{}
+}
+
+// NewNodeLatencyTracker creates a new tracker.
+func NewNodeLatencyTracker() *NodeLatencyTracker {
+	return &NodeLatencyTracker{
+		unneededNodes:    make(map[string]unneededNodeState),
+		unremovableNodes: make(map[string]struct{}),
+	}
+}
+
+// UpdateScaleDownCandidates updates tracked unneeded nodes and reports those that became needed again.
+func (t *NodeLatencyTracker) UpdateScaleDownCandidates(list []*apiv1.Node, timestamp time.Time) {
+	currentSet := make(map[string]struct{}, len(list))
+	for _, node := range list {
+		currentSet[node.Name] = struct{}{}
+		if _, exists := t.unneededNodes[node.Name]; !exists {
+			t.unneededNodes[node.Name] = unneededNodeState{
+				unneededSince: timestamp,
+				threshold:     0,
+			}
+			klog.V(6).Infof("Started tracking unneeded node %s at %v", node.Name, timestamp)
+		}
+	}
+
+	for name, info := range t.unneededNodes {
+		if _, stillUnneeded := currentSet[name]; !stillUnneeded {
+			// This node is no longer a scale-down candidate.
+			if _, isUnremovable := t.unremovableNodes[name]; isUnremovable {
```

Comment:
Interesting, I would expect nodes to first stop being unneeded, before scale down status gets processed. Did you test this?